app/scripts/metamask-controller.js

```diff
@@ -2936,6 +2959,48 @@ export default class MetamaskController extends EventEmitter {
         },
         getAuthorizedScopesByOrigin,
       );
+      if (process.env.MULTICHAIN_API) {
+        this.controllerMessenger.subscribe(
+          `${this.accountsController.name}:selectedAccountChange`,
+          async (account) => {
+            if (
+              account.type === 'solana:data-account' &&
+              account.address !== lastSelectedSolanaAccountAddress
+            ) {
+              lastSelectedSolanaAccountAddress = account.address;
+
+              const originsWithSolanaAccountChangedNotifications =
+                getOriginsWithSessionProperty(
+                  this.permissionController.state,
+                  'solana_accountChanged_notifications',
+                );
+
+              const solanaAccounts = getPermittedAccountsForScopesByOrigin(
+                this.permissionController.state,
+                [
+                  MultichainNetworks.SOLANA,
+                  MultichainNetworks.SOLANA_DEVNET,
+                  MultichainNetworks.SOLANA_TESTNET,
+                ],
+              );
+
+              for (const [origin, accounts] of solanaAccounts.entries()) {
+                const parsedSolanaAddresses = accounts.map((caipAccountId) => {
+                  const { address } = parseCaipAccountId(caipAccountId);
+                  return address;
+                });
+
+                if (
+                  parsedSolanaAddresses.includes(account.address) &&
+                  originsWithSolanaAccountChangedNotifications[origin]
```

Comment:
minor performance optimization if we move this check to the top of this loop

Reply:
🤔 we need the `parsedSolanaAddresses` to be parsed first though?

Reply:
if `originsWithSolanaAccountChangedNotifications[origin]` is false, then we don't need to do `parsedSolanaAddresses` and its includes checks in the first place